eth-bytecode-db/eth-bytecode-db/src/verification/db.rs

```diff
@@ -131,21 +111,116 @@ async fn insert_source_files(
                 .context("insert into \"files\"")?,
             }
         };
+        result.push(file);
+    }
+
+    Ok(result)
+}
+
+async fn insert_source_details(
+    txn: &DatabaseTransaction,
+    source: types::Source,
+    file_models: &[files::Model],
+) -> Result<(sources::Model, bool), anyhow::Error> {
+    let abi = match source.abi {
+        None => None,
+        Some(abi) => serde_json::from_str(&abi).context("deserialize abi")?,
+    };
+
+    let file_ids = {
+        let mut file_ids = file_models.iter().map(|file| file.id).collect::<Vec<_>>();
+        file_ids.sort();
+        file_ids.dedup();
+        file_ids
+    };
+    // Results in `SELECT md5(array_agg(x)::text) FROM (VALUES (1), (2), (3)) t(x).
+    // Without t() `array_agg(x)::text` results in {(1),(2),(3)} instead of {1,2,3}.
+    let file_ids_hash_query = format!(
+        "SELECT md5(array_agg(x)::text)::uuid FROM (VALUES {}) t(x)",
+        file_ids
+            .iter()
+            .map(|i| format!("({i})"))
+            .collect::<Vec<_>>()
+            .join(",")
+    );
+    let file_ids_hash: sea_orm::prelude::Uuid = txn
+        .query_one(Statement::from_string(
+            DatabaseBackend::Postgres,
+            file_ids_hash_query,
+        ))
+        .await
+        .context("calculate hash of file ids")?
+        .ok_or(anyhow::anyhow!(
+            "selection of file ids resulted in empty result"
+        ))
+        .context("calculate hash of file ids")?
+        .try_get("", "md5")
+        .context("calculate hash of file ids")?;
+
+    let (source, inserted) = {
+        let compiler_settings: sea_orm::prelude::Json =
+            serde_json::from_str(&source.compiler_settings)
+                .context("deserialize compiler settings")?;
+
+        let db_source = sources::Entity::find()
+            .filter(sources::Column::CompilerVersion.eq(source.compiler_version.clone()))
+            .filter(sources::Column::CompilerSettings.eq(compiler_settings.clone()))
+            .filter(sources::Column::FileName.eq(source.file_name.clone()))
+            .filter(sources::Column::ContractName.eq(source.contract_name.clone()))
+            .filter(sources::Column::FileIdsHash.eq(file_ids_hash))
+            .one(txn)
+            .await
+            .context("select from \"sources\" by \"compiler_version\", \"compiler_settings\", \"file_name\", \"contract_name\", and \"file_ids_hash\"")?;
 
+        match db_source {
+            Some(source) => (source, false),
+            None => {
+                let source = sources::ActiveModel {
```

Comment:
Wait wait wait
Why do you select first? Is it to check if this exists? This doesn't guarantee you nothing, becausd between select and insert anything can happen
You should make an insert and have some logic ON CONFLICT

Reply:
It may be a problem, I agree. I suggest to fix that in a separate PR, though, as we used that approach before this PR. I suggest to merge those changes and then to fix database data insertion